tests/cpu_bugs/xen_mitigations.pm

```diff
@@ -0,0 +1,349 @@
+                unexpection => {'xl dmesg' => ['^(XEN) *Xen settings: BTI-Thunk JMP, SPEC_CTRL: IBRS+ SSBD-, Other:']}
+            }
+        },
+        "eager-fpu=off" => {
+            "default" => {
+                expection   => {'xl dmesg' => ['Support for HVM VMs: MSR_SPEC_CTRL RSB MD_CLEAR']},
+                unexpection => {'xl dmesg' => ['Support for HVM VMs: MSR_SPEC_CTRL RSB EAGER_FPU MD_CLEAR']}
+            }
+        },
+        "eager-fpu=on" => {
+            "default" => {
+                expection   => {'xl dmesg' => ['Support for HVM VMs: MSR_SPEC_CTRL RSB EAGER_FPU MD_CLEAR']},
+                unexpection => {'xl dmesg' => ['Support for HVM VMs: MSR_SPEC_CTRL RSB MD_CLEAR']}
+            }
+        },
+        "l1d-flush=off" => {
+            "default" => {
+                expection   => {''},
+                unexpection => {'xl dmesg' => ['Xen settings: BTI-Thunk JMP, SPEC_CTRL: IBRS+ SSBD-, Other: .*L1D_FLUSH']},
+            }
+        },
+        "l1d-flush=on" => {
+            "default" => {
+                expection   => {'xl dmesg' => ['Xen settings: BTI-Thunk JMP, SPEC_CTRL: IBRS+ SSBD-, Other: .*L1D_FLUSH']},
+                unexpection => {''},
+            }
+        },
+    },
+  };
+sub do_check {
+    my $secnario = shift;
+    my $foo      = $secnario->{'default'};
+    if ($foo->{'expection'}) {
+        while (my ($cmd, $lines) = each %{$foo->{'expection'}}) {
+            foreach my $l (@{$lines}) {
+                if ($l ne "") {
+                    assert_script_run("$cmd | grep \"$l\"");
+                } else {
+                    print "This expection is empty string, skip";
+                }
+
+            }
+        }
+    }
+    if ($foo->{'unexpection'}) {
+        while (my ($cmd, $lines) = each %{$foo->{'unexpection'}}) {
+            foreach my $l (@{$lines}) {
+                if ($l ne "") {
+                    my $ret = script_run("$cmd | grep \"$l\"");
+                    if ($ret eq 0) {
+                        die "found a unexpection string.";
+                    }
+                } else {
+                    #Debug what output be report.
+                    assert_script_run("xl dmesg | grep -A 10 \"Speculative\"");
+                    print "This unexpection is empty string, skip";
+
+                }
+
+            }
+        }
+    }
+}
+
+sub do_test {
+    my ($self, $hash) = @_;
+    #xen parameter be store into arg
+    #
+    while (my ($arg, $dict) = each %$hash) {
+        while (my ($key, $value) = each %$dict) {
+            my $parameter = $arg . '=' . $key;
+            bootloader_setup::add_grub_xen_cmdline_settings($parameter);
+            bootloader_setup::grub_mkconfig();
+            Mitigation::reboot_and_wait($self, 150);
+            #$value include the check rules of current $parameter.
+            my $ret = do_check($value);
+            if ($ret ne 0) {
```

Comment:
you will never get here , you already dying inside `dio_check` , but I would actually suggest to remove **both** dies here and inside , return 0 from `do_check` and just do `record_info` with result fail . in this way you will run all checks always even if some of them are failing

Reply:
Ok. I will review all of the comments and send a new version.